### PR TITLE
Switch to a non-memory-retaining API for password testing

### PR DESCRIPTION
**Pull request by toolmantim**
TLDR;

* Loads the dictionaries and adjacency graph on demand (reducing retained memory)
* Adds a `Zxcvbn::Tester` for performance-sensitive use cases (for old behaviour)
* Changes to `add_word_lists(key=>val)` from `add_word_list(key,val)`

At the moment we're loading all the dictionary and adjacency graph data when the library is required, and storing it in a constant. This makes for slow load times for Rails applications, and a huge chunk of retained memory. For our Rails application it's the biggest retained memory by gem after ActiveSupport.

The API stays the same:

```ruby
Zxcvbn::Tester.test("password")
```

The change is that `Zxcvbn.test` now loads the dictionary on demand, and doesn't retain it in memory afterwards. This has the advantage of not adding overhead to applications when they add `gem "zxcvbn"` to their Gemfile, and that the base memory usage for the application is much lower. The disadvantage is that is will take longer to test an individual password.

For performance sensitive use cases where retained memory is not an issue, I've added a `Zxcvbn::Tester` class that you can instantiate once (on application initialisation if you wished), for example:

```ruby
tester = Zxcvbn::Tester.new
tester.add_word_lists("custom" => ["words"])
tester.test("password 1")
tester.test("password 2")
tester.test("password 3")
```

For the basic case of `Zxcvbn.test("password")` it will simply load the dictionaries, test the password, and release the memory.

Before this change:

* require allocates 264,315 objects (21MB of memory), retains 87120 objects (6.8MB of memory)
* password test doesn't allocate many objects or retain memory (all happens on require)

```
$ time ruby -I./lib -e ''
real	0m0.176s
user	0m0.090s
sys	0m0.081s

$ time ruby -I./lib -e 'require "zxcvbn"'
real	0m0.325s
user	0m0.219s
sys	0m0.100s

$ ruby -I./lib -rmemory_profiler -e 'puts MemoryProfiler.report { require "zxcvbn" }.pretty_print'
Total allocated 264315
Total retained 87120

allocated memory by gem
-----------------------------------
zxcvbn-ruby/lib x 21126753
rubygems x 602449
json-1.8.2 x 348926
2.2.0/lib x 34868
2.2.0 x 18275
other x 80

$ time ruby -I./lib -rzxcvbn -e 'Zxcvbn.test("password")'
real	0m0.329s
user	0m0.221s
sys	0m0.101s

$ ruby -I./lib -rmemory_profiler -e 'puts MemoryProfiler.report { require "zxcvbn"; Zxcvbn.test("password") }.pretty_print'
Total allocated 265702
Total retained 87123

allocated memory by gem
-----------------------------------
zxcvbn-ruby/lib x 21198553
rubygems x 602369
json-1.8.2 x 348926
2.2.0/lib x 77508
2.2.0 x 18275
other x 160

retained memory by gem
-----------------------------------
zxcvbn-ruby/lib x 6813690
json-1.8.2 x 261837
rubygems x 57182
2.2.0 x 4921
2.2.0/lib x 1649

retained objects by gem
-----------------------------------
zxcvbn-ruby/lib x 85166
json-1.8.2 x 1261
rubygems x 637
2.2.0 x 42
2.2.0/lib x 19
```

After this change:

* require allocates 127 objects (0.01MB of memory), retains 0.006MB
* password test allocates 256,268 objects (21MB of memory), retains 79 objects (0.006MB of memory)

```
$ time ruby -I./lib -e ''
real	0m0.176s
user	0m0.090s
sys	0m0.081s

$ time ruby -I./lib -e 'require "zxcvbn"'
real	0m0.214s
user	0m0.121s
sys	0m0.090s

$ ruby -I./lib -rmemory_profiler -e 'puts MemoryProfiler.report { require "zxcvbn"; }.pretty_print'
Total allocated 7579
Total retained 806

allocated memory by gem
-----------------------------------
rubygems x 621383
json-1.8.2 x 70463
2.2.0 x 18275
2.2.0/lib x 14985
zxcvbn-ruby/lib x 10570
other x 80

allocated objects by gem
-----------------------------------
rubygems x 6233
json-1.8.2 x 878
2.2.0 x 178
2.2.0/lib x 162
zxcvbn-ruby/lib x 127
other x 1

retained memory by gem
-----------------------------------
rubygems x 59752
zxcvbn-ruby/lib x 6490
2.2.0 x 4921
json-1.8.2 x 1214
2.2.0/lib x 880

$ time ruby -I./lib -rzxcvbn -e 'Zxcvbn.test("password")'
real	0m0.324s
user	0m0.221s
sys	0m0.098s

$ ruby -I./lib -rmemory_profiler -e 'puts MemoryProfiler.report { require "zxcvbn"; Zxcvbn.test("password") }.pretty_print'
Total allocated 266098
Total retained 803

allocated memory by gem
-----------------------------------
zxcvbn-ruby/lib x 21200073
rubygems x 621303
json-1.8.2 x 348926
2.2.0/lib x 91801
2.2.0 x 18275
other x 160

allocated objects by gem
-----------------------------------
zxcvbn-ruby/lib x 256268
rubygems x 6232
json-1.8.2 x 2348
2.2.0/lib x 1070
2.2.0 x 178
other x 2

retained memory by gem
-----------------------------------
rubygems x 58070
zxcvbn-ruby/lib x 6730
2.2.0 x 4921
json-1.8.2 x 1841
2.2.0/lib x 1360

retained objects by gem
-----------------------------------
rubygems x 650
zxcvbn-ruby/lib x 79
2.2.0 x 42
2.2.0/lib x 17
json-1.8.2 x 17
```

This is probably a good time to bump to v1.0 as well?

### CODE REVIEW

**Comment by warrenseen** [comment]
These changes look good to me. 

Ping @stevehodgkiss...

**Comment by toolmantim** [comment]
We've been running it in production for a few weeks. Did you want to see a before and after memory footprint for our app?

Seeing as most people aren't "password-strength-as-a-service" and their bottlenecks lie elsewhere I think the default is better to be reduced memory at the price of slightly slower slower performance.

**Comment by toolmantim** [comment]
And if you are "password-strength-as-a-service" you can always:

```ruby
module MyApplication
  PASSWORD_TESTER = Zxcvbn::Tester.new
end

MyApplication::PASSWORD_TESTER.test("password 1")
MyApplication::PASSWORD_TESTER.test("password 2")
MyApplication::PASSWORD_TESTER.test("password 3")
```

**Comment by warrenseen** [comment]
@toolmantim we :heart: numbers, so before/after stats would be cool :smile:

**Comment by stevehodgkiss** [comment]
Hey Tim, thanks for the PR!

The code looks good - we should definitely support both options. I'm not sure which should be the default though... there is a big performance difference between the 2 options - 0.002/password vs 0.095/password on my machine. https://gist.github.com/stevehodgkiss/0183a4f4880472366ea9

**Comment by toolmantim** [comment]
To be clear my main optimisation here was the performance of requiring the gem (allocating memory, and loading the YML), and memory overhead… it wasn't optimising the speed of testing a password. The benchmark was the memory footprint of the app that used it, and the time to `require "zxcvbn"`.

For us, on a daily basis across all our systems and apps we're doing many more fresh `require "zxcvbn"` calls than we are `Zxcvbn.test` calls, and for processes like Sidekiq, cron, etc that will never even invoke Zxcvbn they get a memory reduction and a load speed boost.

I used this similar pattern with the [UserAgentParser](https://github.com/ua-parser/uap-ruby) gem, and just let people load it and assign it to a constant if they wanted… rather than perform lots of expensive stuff at require time.

**Comment by stevehodgkiss** [comment]
Yep, I see where you're coming from... I guess we just need to make it easy for users to choose their tradeoff.

Do you prefer the manual approach over something built into the gem to preload or not?

```ruby
# where sidekiq? indicates whether this process is for running sidekiq...
PASSWORD_TESTER = sidekiq? ? Zxcvbn : Zxcvbn::Tester.new
# vs
Zxcvbn.preload! if sidekiq?
```

**Comment by toolmantim** [comment]
For a typical Rails application I wouldn't suggest to preload it at all… keep the memory for other things and let the GC do it's thing (especially 2.2 and the way Ruby GC is heading). I'm not retaining it in our app (we're on 2.2).

Personally I'm happy with "By default it's threadsafe, doesn't retain memory, etc" along with a Readme and code doc saying "If you're calling it multiple times you can keep the dictionaries around in memory like so" as a Readme/code-doc addition (also, that's why I put such a large comment at the top of the `Zxcvbn::Tester` class).

Before this change:

```
$ ruby -rmemory_profiler -I. -e 'MemoryProfiler.report { require "config/environment" }.pretty_print'
Total allocated 2312988
Total retained 476229

allocated memory by gem
-----------------------------------
activesupport-4.2.1 x 60094221
sprockets-3.0.1 x 56626099
2.2.0/lib x 28225732
zxcvbn-ruby-0.0.3 x 21126729
rubygems x 13918784
json-1.8.2 x 9353860
mime-types-2.4.3 x 7098487
erubis-2.7.0 x 5427205
actionpack-4.2.1 x 5290384
parser-2.2.2.1 x 5268090
...
retained memory by gem
-----------------------------------
sprockets-3.0.1 x 14024689
2.2.0/lib x 7369496
zxcvbn-ruby-0.0.3 x 6813346
activesupport-4.2.1 x 5282673
json-1.8.2 x 3402509
...
retained objects by gem
-----------------------------------
sprockets-3.0.1 x 132076
2.2.0/lib x 85716
zxcvbn-ruby-0.0.3 x 85161
activesupport-4.2.1 x 57325
json-1.8.2 x 22427
mime-types-2.4.3 x 21063
```

After this change:

```
$ ruby -rmemory_profiler -I. -e 'MemoryProfiler.report { require "config/environment" }.pretty_print'
Total allocated 2057572
Total retained 390073

allocated memory by gem
-----------------------------------
activesupport-4.2.1 x 60149050
sprockets-3.0.1 x 56626099
2.2.0/lib x 28229776
rubygems x 13925583
json-1.8.2 x 9075397
mime-types-2.4.3 x 7098487
erubis-2.7.0 x 5427205
actionpack-4.2.1 x 5290384
parser-2.2.2.1 x 5268090
activerecord-4.2.1 x 4941444
...
retained memory by gem
-----------------------------------
sprockets-3.0.1 x 14024689
2.2.0/lib x 7370079
activesupport-4.2.1 x 5285833
mime-types-2.4.3 x 3216794
json-1.8.2 x 3141886
rubygems x 1008626
...
retained objects by gem
-----------------------------------
sprockets-3.0.1 x 132076
2.2.0/lib x 85720
activesupport-4.2.1 x 57352
json-1.8.2 x 21180
mime-types-2.4.3 x 21063
rubygems x 10237
actionpack-4.2.1 x 7414
```

Perhaps @schneems or @samsaffron have a suggestion too?

**Comment by stevehodgkiss** [comment]
I agree, let's conserve memory by default and add a note to the readme about the tradeoff and how to optimise further for performance.

**Comment by toolmantim** [comment]
Cool, I'll update the Readme with the note etc.

**Comment by toolmantim** [comment]
Have updated the Readme with an example

**Comment by toolmantim** [comment]
@stevehodgkiss :+1: :tada:

**Comment by stevehodgkiss** [comment]
Release 0.1.0 with these changes. Thanks @toolmantim!

**Comment by toolmantim** [comment]
@stevehodgkiss it's been 2.5 years. Reckon it's time for a 1.0? ;)

**Comment by stevehodgkiss** [comment]
@toolmantim maybe someday :)
